app/knights/knight.py

```diff
@@ -0,0 +1,40 @@
+class Knight:
+    def __init__(self, knight: dict):
+        self.name = knight["name"]
+        self.power = knight["power"]
+        self.hp = knight["hp"]
+        self.armour = knight["armour"]
+        self.weapon = knight["weapon"]
+        self.potion = knight["potion"]
+        self.protection = 0
+
+    def battle_preparation(self):
+        # apply armour
+        stats = ["power", "protection", "hp"]
+        for armour in self.armour:
+            self.protection += armour["protection"]
+
+        # apply weapon
+        self.power += self.weapon["power"]
+
+        # apply potion if exist
+        if self.potion is not None:
+            for stat in stats:
+                if stat in self.potion["effect"]:
+                    if stat == "power":
+                        self.power += self.potion["effect"]["power"]
+                    if stat == "protection":
+                        self.protection += self.potion["effect"]["protection"]
+                    if stat == "hp":
+                        self.hp += self.potion["effect"]["hp"]
```

Comment:
it also did not resolve code duplication
Try to use this:
`setattr(self, stat, getattr(self, stat) + self.potion["effect"[stat])`